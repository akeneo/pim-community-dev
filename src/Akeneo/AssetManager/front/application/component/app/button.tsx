import * as React from 'react';
import styled from 'styled-components';
import {ThemedProps} from 'akeneoassetmanager/application/component/app/theme';
import Down from 'akeneoassetmanager/application/component/app/icon/down';
import {akeneoTheme} from 'akeneoassetmanager/application/component/app/theme';
import {useShortcut} from 'akeneoassetmanager/application/hooks/input';
import Key from 'akeneoassetmanager/tools/key';

type ButtonProps = {
  buttonSize?: 'micro' | 'medium' | 'default';
  color: 'green' | 'blue' | 'red' | 'grey' | 'outline';
  isDisabled?: boolean;
};

export const ButtonContainer = styled.div`
  display: flex;
  > :not(:first-child) {
    margin-left: 10px;
  }
`;

export const TransparentButton = styled.button`
  background: none;
  border: none;
  padding: 0;
  margin: 0;

  &:hover {
    cursor: pointer;
  }
`;

export const Button = React.forwardRef((props: ButtonProps & any, ref) => (
  <StyledButton ref={ref} {...props} onClick={props.isDisabled ? undefined : props.onClick} />
));

const StyledButton = styled.div<ButtonProps>`
  text-align: center;
  cursor: pointer;
  text-transform: uppercase;
  white-space: nowrap;

  ${(props: ThemedProps<ButtonProps>) => {
    switch (props.buttonSize) {
      case 'micro':
        return `
          padding: 0 10px;
          height: 20px;
          line-height: 19px;
          border-radius: 10px;
          font-size: ${props.theme.fontSize.small};
          min-width: 60px;
        `;
      case 'medium':
        return `
          padding: 0 15px;
          height: 24px;
          line-height: 23px;
          border-radius: 16px;
          font-size: ${props.theme.fontSize.default};
          min-width: 70px;
        `;
      default:
        return `
          padding: 0 15px;
          height: 32px;
          line-height: 30px;
          border-radius: 16px;
          font-size: ${props.theme.fontSize.default};
          min-width: 70px;
        `;
    }
  }}

  ${(props: ThemedProps<ButtonProps>) => {
    if ('outline' === props.color) {
      return `
        color: ${props.theme.color.grey120};
        background-color: white;
        border: 1px solid ${props.theme.color.grey80};
        `;
    }

    return `
      color: white;
      background-color: ${(props.theme.color as any)[props.color + '100']}
      border: 1px solid transparent;
    `;
  }}

  ${(props: ThemedProps<ButtonProps>) =>
    props.isDisabled &&
    `
      cursor: not-allowed;
      opacity: 0.5;
  `}
`;

const Backdrop = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
`;

const Panel = styled.div`
  position: absolute;
  top: 32px;
  right: 0;
  background: white;
  display: flex;
  flex-direction: column;
  box-shadow: 0px 0px 4px 0px rgba(0, 0, 0, 0.3);
  padding: 10px 20px;
  min-width: 180px;
`;

const Container = styled.div`
  position: relative;
`;

const StyledMultipleButton = styled(Button)`
  display: flex;
`;

const Item = styled.div<{isDisabled?: boolean}>`
  color: ${(props: ThemedProps<void>) => props.theme.color.grey120};
  font-size: ${(props: ThemedProps<void>) => props.theme.fontSize.default};
  text-transform: none;
  text-align: left;
  height: 34px;
  display: flex;
  align-items: center;
  cursor: ${props => (props.isDisabled ? 'not-allowed' : 'pointer')};
  opacity: ${props => (props.isDisabled ? 0.5 : 1)};
`;

const DownButton = styled.span`
  display: flex;
  align-items: center;
  padding-left: 15px;
`;

type Item = {
  label: string;
  title?: string;
  isDisabled?: boolean;
  action: () => void;
};

type MultipleButtonProps = {
  items: Item[];
  children: React.ReactNode;
} & ButtonProps;

export const MultipleButton = ({items, children, ...props}: MultipleButtonProps) => {
  const [isOpen, setOpen] = React.useState(false);
  useShortcut(Key.Escape, () => setOpen(false));

  if (0 === items.length) return null;

  const onItemClick = (item: Item) => {
    if (!item.isDisabled) {
      setOpen(false);
      item.action();
    }
  };

  return (
    <Container>
      {1 < items.length ? (
        <>
          <StyledMultipleButton {...props} onClick={() => setOpen(true)}>
            <span>{children}</span>
            <DownButton>
              <Down size={18} color={akeneoTheme.color.white} />
            </DownButton>
          </StyledMultipleButton>
          {isOpen && (
            <>
              <Backdrop onClick={() => setOpen(false)} />
              <Panel>
                {items.map(item => (
                  <Item
                    key={item.label}
                    title={item.title || item.label}
                    isDisabled={item.isDisabled}
                    onClick={() => onItemClick(item)}
                  >
                    {item.label}
                  </Item>
                ))}
              </Panel>
            </>
          )}
        </>
      ) : (
        <Button {...props} onClick={items[0].action}>
          {items[0].label}
        </Button>
      )}
    </Container>
  );
};
