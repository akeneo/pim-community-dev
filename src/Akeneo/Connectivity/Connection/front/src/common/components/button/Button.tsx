import React, {ButtonHTMLAttributes, DetailedHTMLProps, forwardRef, ReactNode, Ref} from 'react';

export type Props = DetailedHTMLProps<ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement> & {
    children?: ReactNode;
    count?: number;
    classNames?: string[];
};

export const Button = forwardRef(({children, count, classNames = [], ...props}: Props, ref: Ref<HTMLButtonElement>) => {
    classNames.push('AknButton');
    if (props.disabled) {
        classNames.push('AknButton--disabled');
    }

    return (
        <button type='button' {...(props as any)} ref={ref} className={`${props.className} ${classNames.join(' ')}`}>
            {children}
            {undefined !== count && <span className='AknButton--withSuffix'>{count}</span>}
        </button>
    );
});
