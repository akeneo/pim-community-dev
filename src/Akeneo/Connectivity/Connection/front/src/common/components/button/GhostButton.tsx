import styled from '../../styled-with-theme';

export const GhostButton = styled.button`
    background-color: white;
    border-radius: 16px;
    border: 1px solid ${({theme}) => theme.color.grey100};
    color: ${({theme}) => theme.color.grey120};
    cursor: pointer;
    font-size: ${({theme}) => theme.fontSize.big};
    height: 32px;
    line-height: 30px;
    min-width: 32px;
    padding: 0 16px;
    text-transform: uppercase;
    transition: all 0.1s ease;
    white-space: nowrap;

    :hover {
        background-color: ${({theme}) => theme.color.grey20};
        color: ${({theme}) => theme.color.grey140};
    }

    :disabled {
        border: 1px solid ${({theme}) => theme.color.grey60};
        color: ${({theme}) => theme.color.grey80};
        cursor: not-allowed;
    }

    :focus {
        box-shadow: 0px 0px 2px 0px ${({theme}) => theme.color.blue100};
        outline: none;
    }

    :active {
        border: 1px solid ${({theme}) => theme.color.grey140};
        color: ${({theme}) => theme.color.grey140};
    }
`;
