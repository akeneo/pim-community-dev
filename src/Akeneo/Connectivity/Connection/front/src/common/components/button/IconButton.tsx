import styled from 'styled-components';
import {getColor} from 'akeneo-design-system';

export const IconButton = styled.button`
    background-color: transparent;
    border: none;
    cursor: pointer;
    display: inline-flex;
    height: 24px;
    justify-content: center;
    opacity: 0.8;
    padding: 0;
    transition: opacity 0.1s ease-in;
    width: 24px;
    color: ${getColor('grey', 100)};

    :hover {
        opacity: 1;
    }

    :focus {
        outline: none;
    }

    svg {
        width: 18px;
        height: 18px;
    }
`;
