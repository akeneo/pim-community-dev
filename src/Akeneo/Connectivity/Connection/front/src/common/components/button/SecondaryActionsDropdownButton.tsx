import React, {ReactNode, DetailedHTMLProps, ButtonHTMLAttributes} from 'react';
import {Translate} from '../../../shared/translate';

export const SecondaryActionsDropdownButton = ({children}: {children: ReactNode}) => (
    <div className='AknSecondaryActions AknDropdown AknButtonList-item'>
        <div className='AknSecondaryActions-button' data-toggle='dropdown'></div>
        <div className='AknDropdown-menu AknDropdown-menu--right'>
            <div className='AknDropdown-menuTitle'>
                <Translate id='akeneo_connectivity.connection.secondary_actions.title' />
            </div>
            {children}
        </div>
    </div>
);

export const DropdownLink = ({
    children,
    ...props
}: DetailedHTMLProps<ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement>) => (
    <button type='button' className='AknDropdown-menuLink' {...props}>
        {children}
    </button>
);
