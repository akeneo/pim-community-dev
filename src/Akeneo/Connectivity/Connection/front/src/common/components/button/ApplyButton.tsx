import React, {forwardRef, Ref} from 'react';
import {Button, Props} from './Button';

export const ApplyButton = forwardRef(({classNames = [], ...props}: Props, ref: Ref<HTMLButtonElement>) => {
    classNames.push('AknButton--apply');

    return <Button {...props} ref={ref} classNames={classNames}></Button>;
});
