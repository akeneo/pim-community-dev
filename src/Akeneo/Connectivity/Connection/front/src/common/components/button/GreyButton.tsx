import React, {forwardRef, Ref} from 'react';
import {Button, Props} from './Button';

export const GreyButton = forwardRef(({classNames = [], ...props}: Props, ref: Ref<HTMLButtonElement>) => {
    classNames.push('AknButton--grey');

    return <Button {...props} ref={ref} classNames={classNames}></Button>;
});
