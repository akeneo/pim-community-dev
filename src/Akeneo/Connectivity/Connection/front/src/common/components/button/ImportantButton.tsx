import React, {forwardRef, Ref} from 'react';
import {Button, Props} from './Button';

export const ImportantButton = forwardRef(({classNames = [], ...props}: Props, ref: Ref<HTMLButtonElement>) => {
    classNames.push('AknButton--important');

    return <Button {...props} ref={ref} classNames={classNames}></Button>;
});
