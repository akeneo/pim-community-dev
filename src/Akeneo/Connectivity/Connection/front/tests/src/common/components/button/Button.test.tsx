import React from 'react';
import {create} from 'react-test-renderer';
import {Button} from '@src/common/components/button/Button';

describe('Button', () => {
    it('should render', () => {
        const component = create(<Button onClick={() => undefined}>content</Button>);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it('should be disabled', () => {
        const component = create(
            <Button onClick={() => undefined} disabled>
                content
            </Button>
        );

        expect(component.toJSON()).toMatchSnapshot();
    });

    it('should display a counter', () => {
        const component = create(
            <Button onClick={() => undefined} count={9}>
                content
            </Button>
        );

        expect(component.toJSON()).toMatchSnapshot();
    });

    it('should have custom classes', () => {
        const component = create(
            <Button onClick={() => undefined} classNames={['class1', 'class2']}>
                content
            </Button>
        );

        expect(component.toJSON()).toMatchSnapshot();
    });
});
