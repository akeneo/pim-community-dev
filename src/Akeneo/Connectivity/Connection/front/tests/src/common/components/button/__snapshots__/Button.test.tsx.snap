// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should be disabled 1`] = `
<button
  className="AknButton AknButton--disabled"
  disabled={true}
  onClick={[Function]}
  type="button"
>
  content
</button>
`;

exports[`Button should display a counter 1`] = `
<button
  className="AknButton"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  content
  <span
    className="AknButton--withSuffix"
  >
    9
  </span>
</button>
`;

exports[`Button should have custom classes 1`] = `
<button
  className="class1 class2 AknButton"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  content
</button>
`;

exports[`Button should render 1`] = `
<button
  className="AknButton"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  content
</button>
`;
